utils/src/lib.rs

```diff
@@ -62,9 +65,43 @@ pub fn remove_extra_costs(gas: U256, calldata: &[u8]) -> U256 {
     gas - calldata_cost - BASE_TX_COST
 }
 
+/// Flattens a group of contracts into maps of all events and functions
+pub fn flatten_known_contracts(
+    contracts: &BTreeMap<String, (Abi, Vec<u8>)>,
+) -> (BTreeMap<[u8; 4], Function>, BTreeMap<H256, Event>, Abi) {
+    let flattened_funcs: BTreeMap<[u8; 4], Function> = contracts
+        .iter()
+        .flat_map(|(_name, (abi, _code))| {
+            abi.functions()
+                .map(|func| (func.short_signature(), func.clone()))
+                .collect::<BTreeMap<[u8; 4], Function>>()
+        })
+        .collect();
+
+    let flattened_events: BTreeMap<H256, Event> = contracts
+        .iter()
+        .flat_map(|(_name, (abi, _code))| {
+            abi.events()
+                .map(|event| (event.signature(), event.clone()))
+                .collect::<BTreeMap<H256, Event>>()
```

Comment:
this is probably overhead and can be removed

Reply:
feel free to suggest an alternative - i have rewritten this a few times now

Reply:
Ima take a stab at this in the morning thanks for dealing with the nits 😅